Data Retention Policy API Docs (#556)
hashicorp/terraform-docs-common: website/docs/cloud-docs/api-docs/workspaces.mdx

Status code `204`.

> ## Show data retention policy

Reviewer: Non-Blocking Nitpick (Same as above!): Should we titlecase this for consistency between the other headers and this one?:
```suggestion
## Show Data Retention Policy
```

This endpoint shows the data retention policy set explicitly on the workspace.
When no data retention policy is set on the workspace, the organization data retention policy will be used by default. Read more about [workspace data retention policies](/terraform/cloud-docs/workspaces/settings/deletion#data-retention-policies).

For more information on the types of data retention policies, refer to [Data Retention Policy Types](/terraform/cloud-docs/api-docs/data-retention-policies#types-of-data-retention-policies)

| Status  | Response                                        | Reason                                                        |
| ------- | ----------------------------------------------- | ------------------------------------------------------------- |
| [200][] | [JSON API document][] (`type: "data-retention-policy-delete-olders" or "data-retention-policy-dont-deletes"`) | The request was successful. [Types of Data Retention Policies](/terraform/cloud-docs/api-docs/data-retention-policies#types-of-data-retention-policies). |

> | [404][] | [JSON API error object][]                       | Workspace not found, data retention policy does not exist, or user unauthorized to perform action |

Reviewer: Non-blocking Nitpick (same as above!): Should we add a period here for consistency?:
```suggestion
| [404][] | [JSON API error object][]                       | Workspace not found, data retention policy does not exist, or user unauthorized to perform action. |
```

### Request Body

No request body.

### Sample Request

```shell
$ curl \
  --header "Authorization: Bearer $TOKEN" \
  --header "Content-Type: application/vnd.api+json" \
  https://app.terraform.io/api/v2/workspaces/workspace-2/relationships/data-retention-policy
```

### Sample Response

```json
{
  "data": {
    "id": "drp-Kh74zpKVGNWbpugQ",
    "type": "data-retention-policy-delete-olders",
    "attributes": {
      "delete-older-than-n-days": 60
    },
    "relationships": {
      "target": {
        "data": {
          "id": "workspace-2",
          "type": "workspaces"
        }
      }
    }
  }
}
```

> ## Create or update data retention policy

Reviewer: Non-Blocking Nitpick (Same as above!): Should we titlecase this for consistency between the other headers and this one?:
```suggestion
## Create or Update Data Retention Policy
```

| Parameter       | Description                                                                                                                                                         |
| --------------- | ------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
| `:workspace_id` | The workspace ID to update the data retention policy for. Obtain this from the [workspace settings](/terraform/cloud-docs/workspaces/settings) or the [Show Workspace](#show-workspace) endpoint. |

| Status  | Response                  | Reason(s)                                                   |
| ------- | ------------------------- | ----------------------------------------------------------- |
| [204][] | No Content                | Successfully updated the workspace data retention policy.   |

> | [404][] | [JSON API error object][] | Workspace not found, or user unauthorized to perform action |

Reviewer: Non-blocking Nitpick (same as above!): Should we add a period here for consistency?:
```suggestion
| [404][] | [JSON API error object][] | Workspace not found, or user unauthorized to perform action. |
```

### Request Body

> This POST endpoint requires a JSON object with the following properties as a request payload.

Reviewer: Suggestion (same as above!): Should we format this for consistency?:
```suggestion
This `POST` endpoint requires a JSON object with the following properties as a request payload.
```

| `data.type`       | string |         | A [data retention policy type](/terraform/cloud-docs/api-docs/data-retention-policies#types-of-data-retention-policies) |
| `data.attributes` | object |         | The attributes for the [chosen policy type](/terraform/cloud-docs/api-docs/data-retention-policies#types-of-data-retention-policies) |

### Sample Payload

```json
{
  "data": {
    "type": "data-retention-policy-delete-olders",
    "attributes": {
      "deleteOlderThanNDays": 33
    }
  }
}
```

### Sample Request

```shell
$ curl \
  --header "Authorization: Bearer $TOKEN" \
  --header "Content-Type: application/vnd.api+json" \
  --request POST \
  --data @payload.json \
  https://app.terraform.io/api/v2/workspaces/ws-UYv6RYM8fVhzeGG5/relationships/data-retention-policy
```

### Sample Response

```json
{
  "data": {
    "id": "drp-Kh74zpKVGNWbpugQ",
    "type": "data-retention-policy-delete-olders",
    "attributes": {
      "delete-older-than-n-days": 33
    },
    "relationships": {
      "target": {
        "data": {
          "id": "workspace-2",
          "type": "workspaces"
        }
      }
    }
  }
}
```

> ## Remove data retention policy

Reviewer: Non-Blocking Nitpick (Same as above!): Should we titlecase this for consistency between the other headers and this one?:
```suggestion
## Remove Data Retention Policy
```

<EnterpriseAlert>
This endpoint is exclusive to Terraform Enterprise, and not available in Terraform Cloud.
</EnterpriseAlert>

`DELETE /workspaces/:workspace_id/relationships/data-retention-policy`

| Parameter       | Description                                                                                                                                                                          |
| --------------- | ------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------ |
| `:workspace_id` | The workspace ID to remove the data retenetion policy for. Obtain this from the [workspace settings](/terraform/cloud-docs/workspaces/settings) or the [Show Workspace](#show-workspace) endpoint. |

This endpoint removes the data retention policy explicitly set on a workspace.

> Workspaces without a data retention policy will default to using the organization's data retention policy when determining how long to retain backing data.

Reviewer: Non-Blocking Suggestion: Would it be beneficial to link the org DRP docs here for easy access?:
```suggestion
Workspaces without a data retention policy will default to using the [organization's data retention policy](https://developer.hashicorp.com/terraform/enterprise/users-teams-organizations/organizations#destruction-and-deletion) when determining how long to retain backing data.
```

> Workspaces without a data retention policy will default to using the [organization's data retention policy](/terraform/enterprise/users-teams-organizations/organizations#destruction-and-deletion) when determining how long to retain backing data.

Reviewer 2: ```suggestion
When no data retention policy is set for the workspace, the endpoint returns the default policy configured for the organization. Refer to [Data Retention Policies](/terraform/enterprise/users-teams-organizations/organizations#destruction-and-deletion) for instructions on configuring data retention policies for organizations.
```

Read more about [workspace data retention policies](/terraform/cloud-docs/workspaces/settings/deletion#data-retention-policies).

| Status  | Response                  | Reason(s)                                                             |
| ------- | ------------------------- | --------------------------------------------------------------------- |
| [204][] | No Content                | Successfully removed the workspace data retention policy              |

> | [404][] | [JSON API error object][] | Workspace not found, or user unauthorized to perform action           |

Reviewer 2: ```suggestion
| [204][] | No content                | Successfully removed the workspace data retention policy              |
| [404][] | [JSON API error object][] | Workspace not found or user is not authorized to perform action           |
```

### Request Body

No request body.

### Sample Request

```shell
$ curl \
  --header "Authorization: Bearer $TOKEN" \
  --header "Content-Type: application/vnd.api+json" \
  --request DELETE \
  https://app.terraform.io/api/v2/workspaces/ws-UYv6RYM8fVhzeGG5/relationships/data-retention-policy
```

### Response

No response body.

Status code `204`.

> 

Reviewer 2: ```suggestion
<!-- END:   TFE:only -->
```

Status code `204`.

## Show data retention policy

<EnterpriseAlert>

> This endpoint is exclusive to Terraform Enterprise, and not available in Terraform Cloud.

Reviewer 2: ```suggestion
This endpoint is exclusive to Terraform Enterprise and is not available in Terraform Cloud.
```

</EnterpriseAlert>

`GET /workspaces/:workspace_id/relationships/data-retention-policy`

| Parameter            | Description                                                         |
| ---------------------| --------------------------------------------------------------------|
| `:workspace_id`      | The ID of the workspace to show the data retention policy for. Obtain this from the [workspace settings](/terraform/cloud-docs/workspaces/settings) or the [Show Workspace](#show-workspace) endpoint. |

This endpoint shows the data retention policy set explicitly on the workspace.

> When no data retention policy is set on the workspace, the organization data retention policy will be used by default. Read more about [workspace data retention policies](/terraform/cloud-docs/workspaces/settings/deletion#data-retention-policies).

Reviewer: Non-blocking nitpick (same as above!): Should we add an extra line between this line of text and the previous line of text to keep it consistent with the [text on line 948](https://github.com/hashicorp/terraform-docs-common/pull/556/files#diff-f5625a4be7f4b34305dd570e3a8eb331c34e7fd7bc58dab8f5499ed46ceb8737R948)?:
```suggestion
When no data retention policy is set on the workspace, the organization data retention policy will be used by default. 

Read more about [workspace data retention policies](/terraform/cloud-docs/workspaces/settings/deletion#data-retention-policies).
```

> See [Data Retention Policy API](/terraform/enterprise/api-docs/data-retention-policies#show-data-retention-policy) for details.

Reviewer 2: ```suggestion
Refer to [Data Retention Policy API](/terraform/enterprise/api-docs/data-retention-policies#show-data-retention-policy) for details.
```

## Create or update data retention policy

<EnterpriseAlert>

> This endpoint is exclusive to Terraform Enterprise, and not available in Terraform Cloud.

Reviewer 2: ```suggestion
This endpoint is exclusive to Terraform Enterprise and is not available in Terraform Cloud.
```

</EnterpriseAlert>

`POST /workspaces/:workspace_id/relationships/data-retention-policy`

| Parameter       | Description                                                                                                                                                         |
| --------------- | ------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
| `:workspace_id` | The workspace ID to update the data retention policy for. Obtain this from the [workspace settings](/terraform/cloud-docs/workspaces/settings) or the [Show Workspace](#show-workspace) endpoint. |

This endpoint creates a data retention policy for a workspace, or updates the existing policy. 

> Read more about [workspace data retention policies](/terraform/cloud-docs/workspaces/settings/deletion#data-retention-policies).

Reviewer: Non-blocking nitpick (same as above!): Should we add an extra line between this line of text and the previous line of text to keep it consistent with the [text on line 948](https://github.com/hashicorp/terraform-docs-common/pull/556/files#diff-f5625a4be7f4b34305dd570e3a8eb331c34e7fd7bc58dab8f5499ed46ceb8737R948)?:
```suggestion

Read more about [workspace data retention policies](/terraform/cloud-docs/workspaces/settings/deletion#data-retention-policies).
```

> See [Data Retention Policy API](/terraform/enterprise/api-docs/data-retention-policies#create-or-update-data-retention-policy) for details.

Reviewer 2: ```suggestion
Refer to [Data Retention Policy API](/terraform/enterprise/api-docs/data-retention-policies#create-or-update-data-retention-policy) for details.
```

## Remove data retention policy

<EnterpriseAlert>

> This endpoint is exclusive to Terraform Enterprise, and not available in Terraform Cloud.

Reviewer 2: ```suggestion
This endpoint is exclusive to Terraform Enterprise and is not available in Terraform Cloud.
```

| Parameter       | Description                                                                                                                                                                          |
| --------------- | ------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------ |
| `:workspace_id` | The workspace ID to remove the data retenetion policy for. Obtain this from the [workspace settings](/terraform/enterprise/workspaces/settings) or by sending a `GET` request to the [`/workspaces`](#show-workspace) endpoint. |

This endpoint removes the data retention policy explicitly set on a workspace.
When no data retention policy is set for the workspace, the endpoint returns the default policy configured for the organization. Refer to [Data Retention Policies](/terraform/enterprise/users-teams-organizations/organizations#destruction-and-deletion) for instructions on configuring data retention policies for organizations.

Read more about [workspace data retention policies](/terraform/enterprise/workspaces/settings/deletion#data-retention-policies).

> See [Data Retention Policy API](/terraform/enterprise/api-docs/data-retention-policies#remove-data-retention-policy) for details.

Reviewer 2: ```suggestion
Refer to [Data Retention Policy API](/terraform/enterprise/api-docs/data-retention-policies#remove-data-retention-policy) for details.
```
